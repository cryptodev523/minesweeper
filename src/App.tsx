import { useEffect, useMemo, useState } from 'react';
import { Typography, Select, Box, MenuItem, SelectChangeEvent } from '@mui/material';
import LoadingButton from '@mui/lab/LoadingButton';

import useStyles from './styles';
import { levels } from './utils';
import * as types from './store/types';
import { useAppDispatch, useAppSelector } from './store/hooks';
import { isLoadingSelector, statusSelector, mapSelector } from './stage/stageSlice';
import Stage from './stage';

function App() {
  const classes = useStyles();
  const dispatch = useAppDispatch();
  const isLoading = useAppSelector(isLoadingSelector);
  const status = useAppSelector(statusSelector);
  const map = useAppSelector(mapSelector);

  const [level, setLevel] = useState<string>('new 1');

  useEffect(() => {
    dispatch({ type: types.INITIALIZE_GAME });
  }, []);

  const message = useMemo(() => {
    switch (status) {
      case 'OK':
        return 'Playing';
      case 'You lose':
        return (
          <>
            You lose.
            <br /> Click START button to restart new game!
          </>
        );
      default:
        return 'Click START button to start new game!';
    }
  }, [status]);

  const handleLevelChange = (e: SelectChangeEvent) => {
    setLevel(e.target.value);
  };

  const handleStart = () => {
    dispatch({ type: types.START_GAME, payload: level });
  };

  return (
    <Box className={classes.container}>
      <Box className={classes.controls}>
        <Typography variant="h3" mt={6} mb={1}>
          Welcome! Enjoy Minesweeper
        </Typography>

        <Box className={classes.selectWrapper}>
          <Select
            className={classes.select}
            variant="standard"
            value={level}
            label="Level"
            onChange={handleLevelChange}
          >
            {levels.map((item) => (
              <MenuItem key={item.value} value={item.value}>
                {item.label}
              </MenuItem>
            ))}
          </Select>
          <LoadingButton
            variant="outlined"
            data-testid="btn-start"
            onClick={handleStart}
            disabled={isLoading}
            loading={isLoading}
          >
            Start
          </LoadingButton>
        </Box>

        <Typography variant="subtitle1" align="center" mt={2} mb={3}>
          {message}
        </Typography>
      </Box>

      <Box className={classes.boardWrapper}>
        <Stage map={map} />
      </Box>
    </Box>
  );
}

export default App;
