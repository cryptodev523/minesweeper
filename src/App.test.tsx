import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { cleanup, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';

import watcher from './store/saga';
import App from './App';

const sagaMiddleware = createSagaMiddleware();
const mockStore = configureStore([sagaMiddleware]);

describe('App', () => {
  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it('should render correctly and match with snapshot', () => {
    const initialState = {
      board: {
        isLoading: false,
        map: [],
        status: '',
      },
    };
    const store = mockStore(initialState);

    const tree = renderer
      .create(
        <Provider store={store}>
          <App />
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should include title, start button, and start guide at first render', () => {
    const initialState = {
      board: {
        isLoading: false,
        map: [],
        status: '',
      },
    };

    const store = mockStore(initialState);

    const wrapper = mount(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    expect(wrapper.text().includes('Minesweeper')).toBe(true);
    expect(wrapper.text().includes('Click START button to start new game!')).toBe(true);
    expect(wrapper.text().includes('Start')).toBe(true);
  });

  it('should disable start button when loading new game', async () => {
    const initialState = {
      board: {
        isLoading: true,
        map: [],
        status: '',
      },
    };
    const store = mockStore(initialState);
    sagaMiddleware.run(watcher);

    const wrapper = render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    expect(wrapper.getByText('Start')).toBeDisabled();
  });

  it('should include playing message after render board', async () => {
    const initialState = {
      board: {
        isLoading: false,
        map: [
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
        ],
        status: 'OK',
      },
    };

    const store = mockStore(initialState);
    sagaMiddleware.run(watcher);

    const wrapper = mount(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    expect(wrapper.text().includes('Playing')).toBe(true);
  });
});
