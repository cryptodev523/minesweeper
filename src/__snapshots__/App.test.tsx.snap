// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly and match with snapshot 1`] = `
<div
  className="makeStyles-container-1 MuiBox-root css-0"
>
  <div
    className="makeStyles-controls-2 MuiBox-root css-0"
  >
    <h3
      className="MuiTypography-root MuiTypography-h3 css-zxqeea-MuiTypography-root"
    >
      Welcome! Enjoy Minesweeper
    </h3>
    <div
      className="makeStyles-selectWrapper-3 MuiBox-root css-0"
    >
      <div
        className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary makeStyles-select-4 css-194ccxm-MuiInputBase-root-MuiInput-root-MuiSelect-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          className="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Level 1
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value="new 1"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <button
        className="MuiLoadingButton-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-6m3mwv-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        data-testid="btn-start"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Start
      </button>
    </div>
    <h6
      className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-alignCenter css-1sqxg7g-MuiTypography-root"
    >
      Click START button to start new game!
    </h6>
  </div>
  <div
    className="makeStyles-boardWrapper-5 MuiBox-root css-0"
  >
    <div
      className="makeStyles-board-6 MuiBox-root css-0"
    />
  </div>
</div>
`;
